In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [6]:
DATA_PATH = "../docs/master_property_tax_enforcement_dataset.csv"

master_property = pd.read_csv(DATA_PATH, low_memory=False)

print(master_property.shape)
display(master_property.head())

(9763, 73)


,month,cycle,borough_lien,block_lien,lot_lien,tax_class_code,building_class,community_board,council_district,house_number,...,vacancy_rate,rent_burden_rate,severe_rent_burden_rate,black_rate,hispanic_rate,asian_rate,white_nonhispanic_rate,age_65_plus_rate,bachelors_plus_rate,matched_acs
0,2019-10-01T00:00:00.000,Final Sale,1,26,1260,2,R4,101.0,1.0,15,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
1,2019-10-01T00:00:00.000,Final Sale,1,26,1273,2,R4,101.0,1.0,15,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
2,2019-10-01T00:00:00.000,Final Sale,1,27,1040,2,R4,101.0,1.0,55,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
3,2019-10-01T00:00:00.000,Final Sale,1,31,1068,2,R4,101.0,1.0,75,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
4,2019-10-01T00:00:00.000,Final Sale,1,44,1265,2,R4,101.0,1.0,18,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0


#### Section 1: Property Composition

What type of properties dominate the lien sale pool?

In [20]:
# TAX CLASS

tax_class_summary = (
    master_property["tax_class_code"]
    .value_counts(dropna=False)
    .rename_axis("tax_class")
    .reset_index(name="properties")
)

tax_class_summary["share"] = (
    tax_class_summary["properties"] /
    tax_class_summary["properties"].sum()
)

display(tax_class_summary)

,tax_class,properties,share
0,1,4744,0.485916
1,4,2718,0.278398
2,2,2301,0.235686


<B>
The final-sale pool is dominated by residential property. </B>

 Final-sale pool is:

48.6% Class 1 <BR>
23.6% Class 2 <br>
27.8% Class 4 <br>


In [21]:

# Borough composition

borough_summary = (
    master_property["borough_lien"]
    .value_counts(dropna=False)
    .rename_axis("borough")
    .reset_index(name="properties")
)

borough_summary["share"] = (
    borough_summary["properties"] /
    borough_summary["properties"].sum()
)

display(borough_summary)

,borough,properties,share
0,3,3715,0.380518
1,4,2845,0.291406
2,2,1324,0.135614
3,1,1067,0.109290
4,5,812,0.083171


<b>
Brooklyn and Queens Dominate the Final-Sale Pool
</B> <BR>

Brooklyn: 38.1% <BR>
Queens: 29.1% <BR>
Bronx: 13.6% <BR>
Manhattan: only 10.9% <BR>

In [22]:
# Building class composition

building_class_summary = (
    master_property["building_class"]
    .value_counts(dropna=False)
    .head(25)
    .rename_axis("building_class")
    .reset_index(name="properties")
)

building_class_summary["share"] = (
    building_class_summary["properties"] /
    building_class_summary["properties"].sum()
)

display(building_class_summary)

,building_class,properties,share
0,C0,796,0.099562
1,R4,633,0.079174
2,B2,630,0.078799
3,B1,619,0.077423
4,V0,615,0.076923
5,B3,498,0.062289
6,RG,420,0.052533
7,A1,398,0.049781
8,C3,352,0.044028
9,C2,327,0.040901


In [23]:
# PLUTO land use composition

landuse_summary = (
    master_property["landuse"]
    .value_counts(dropna=False)
    .rename_axis("landuse")
    .reset_index(name="properties")
)

landuse_summary["share"] = (
    landuse_summary["properties"] /
    landuse_summary["properties"].sum()
)

display(landuse_summary)

,landuse,properties,share
0,1.0,2926,0.299703
1,NaN,2033,0.208235
2,2.0,1876,0.192154
3,4.0,906,0.092799
4,11.0,853,0.087371
5,5.0,393,0.040254
6,10.0,264,0.027041
7,6.0,183,0.018744
8,7.0,130,0.013316
9,8.0,116,0.011882


In [24]:
# Residential unit distribution

master_property["unitsres"] = pd.to_numeric(
    master_property["unitsres"],
    errors="coerce"
)

master_property["unitstotal"] = pd.to_numeric(
    master_property["unitstotal"],
    errors="coerce"
)

display(
    master_property[[
        "unitsres",
        "unitstotal"
    ]].describe()
)

,unitsres,unitstotal
count,7776.000000,7776.000000
mean,3.002443,3.369727
std,9.030559,9.144739
min,0.000000,0.000000
25%,0.000000,1.000000
50%,2.000000,2.000000
75%,3.000000,3.000000
max,355.000000,355.000000


In [25]:
# Small homes vs larger buildings

master_property["small_property_flag"] = np.where(
    master_property["unitsres"].fillna(0) <= 3,
    1,
    0
)

small_property_summary = (
    master_property["small_property_flag"]
    .value_counts(dropna=False)
    .rename_axis("small_property_flag")
    .reset_index(name="properties")
)

small_property_summary["share"] = (
    small_property_summary["properties"] /
    small_property_summary["properties"].sum()
)

display(small_property_summary)

,small_property_flag,properties,share
0,1,8283,0.848407
1,0,1480,0.151593


<b> Small Residential Buildings Dominate </b> <br>

C0/C1/C2/C3/C7 → walk-up apartments <br>
B1/B2/B3 → 2-family homes <br>
A1/A2/A5 → 1-family homes <br>
R4 → condo-related <br>
S2 → mixed-use residential  <br>

Most final-sale properties are very small properties (84%)

In [26]:
# Property value distribution

value_cols = [
    "fullval",
    "avtot",
    "avland",
    "extot"
]

for col in value_cols:
    master_property[col] = pd.to_numeric(
        master_property[col],
        errors="coerce"
    )

display(
    master_property[value_cols].describe()
)

,fullval,avtot,avland,extot
count,9.571000e+03,9.571000e+03,9.571000e+03,9.571000e+03
mean,6.889297e+05,1.278419e+05,3.364386e+04,1.200426e+04
std,1.382257e+06,5.586483e+05,1.210361e+05,3.614989e+05
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.317895e+05,1.846450e+04,4.587500e+03,0.000000e+00
50%,5.090000e+05,3.094100e+04,8.863000e+03,0.000000e+00
75%,7.700000e+05,9.126500e+04,2.205000e+04,0.000000e+00
max,7.532900e+07,3.389805e+07,5.220000e+06,3.389805e+07


Median full market value is $509K. <br>

In [27]:
# Top building classes by HPD distress

building_hpd_summary = (
    master_property
    .groupby("building_class", as_index=False)
    .agg(
        properties=("bbl", "nunique"),
        avg_open_hpd_violations=("open_hpd_violations", "mean"),
        avg_open_c_violations=("open_c_violations", "mean")
    )
    .sort_values(
        "avg_open_hpd_violations",
        ascending=False
    )
)

display(building_hpd_summary.head(20))

,building_class,properties,avg_open_hpd_violations,avg_open_c_violations
21,C9,4,127.500000,62.250000
22,D1,26,76.923077,33.307692
117,S0,3,73.000000,14.000000
28,D7,18,68.833333,24.222222
17,C4,40,66.700000,15.250000
14,C1,279,64.892473,18.630824
20,C7,89,62.730337,18.179775
19,C6,3,58.666667,3.666667
18,C5,71,53.408451,11.014085
24,D3,11,51.909091,19.090909


Older walk-up classes,older multifamily stock,
certain mixed-use classes,
and elevator/walk-up apartments

associated with extremely high violation counts.

## Section 2: Neighborhood vulnerability

In [28]:
# ACS variable overview

acs_cols = [
    "median_household_income",
    "poverty_rate",
    "rent_burden_rate",
    "severe_rent_burden_rate",
    "homeownership_rate",
    "vacancy_rate",
    "age_65_plus_rate",
    "black_rate",
    "hispanic_rate",
    "asian_rate",
    "white_nonhispanic_rate",
    "bachelors_plus_rate"
]

display(master_property[acs_cols].describe())

,median_household_income,poverty_rate,rent_burden_rate,severe_rent_burden_rate,homeownership_rate,vacancy_rate,age_65_plus_rate,black_rate,hispanic_rate,asian_rate,white_nonhispanic_rate,bachelors_plus_rate
count,7.088000e+03,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000,7084.000000
mean,-8.288863e+06,0.173186,0.520578,0.290783,0.392412,0.076447,0.148388,0.323129,0.276480,0.110529,0.228713,0.325374
std,7.424768e+07,0.109520,0.141023,0.122991,0.241981,0.054853,0.062580,0.292335,0.214338,0.145471,0.255722,0.169978
min,-6.666667e+08,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.032719
25%,5.977300e+04,0.088717,0.425957,0.205628,0.195889,0.040289,0.105205,0.042683,0.109194,0.018005,0.022137,0.204649
50%,8.038800e+04,0.153245,0.515592,0.284519,0.350567,0.066308,0.141243,0.248567,0.211538,0.053153,0.119605,0.294118
75%,9.942400e+04,0.234157,0.607710,0.362998,0.577810,0.100542,0.184307,0.574308,0.402370,0.140600,0.372469,0.400188
max,2.500010e+05,0.671744,1.000000,0.743757,0.995338,0.496933,0.556254,0.933546,0.936729,0.897676,0.999027,0.958750


In [29]:

# Tract-level summary dataset

tract_summary = (
    master_property
    .dropna(subset=["geoid"])
    .groupby("geoid", as_index=False)
    .agg(
        final_sale_properties=("bbl", "nunique"),
        avg_open_hpd_violations=("open_hpd_violations", "mean"),
        total_open_hpd_violations=("open_hpd_violations", "sum"),
        avg_open_c_violations=("open_c_violations", "mean"),
        median_household_income=("median_household_income", "first"),
        poverty_rate=("poverty_rate", "first"),
        rent_burden_rate=("rent_burden_rate", "first"),
        severe_rent_burden_rate=("severe_rent_burden_rate", "first"),
        homeownership_rate=("homeownership_rate", "first"),
        vacancy_rate=("vacancy_rate", "first"),
        age_65_plus_rate=("age_65_plus_rate", "first"),
        black_rate=("black_rate", "first"),
        hispanic_rate=("hispanic_rate", "first"),
        asian_rate=("asian_rate", "first"),
        white_nonhispanic_rate=("white_nonhispanic_rate", "first"),
        bachelors_plus_rate=("bachelors_plus_rate", "first")
    )
)

print(tract_summary.shape)
display(tract_summary.head())

(1665, 17)


,geoid,final_sale_properties,avg_open_hpd_violations,total_open_hpd_violations,avg_open_c_violations,median_household_income,poverty_rate,rent_burden_rate,severe_rent_burden_rate,homeownership_rate,vacancy_rate,age_65_plus_rate,black_rate,hispanic_rate,asian_rate,white_nonhispanic_rate,bachelors_plus_rate
0,3.600500e+10,21,7.476190,157.0,0.714286,121171.0,0.140815,0.365979,0.240550,0.597510,0.038564,0.138111,0.293027,0.617732,0.060073,0.014873,0.339491
1,3.600500e+10,7,0.000000,0.0,0.000000,98242.0,0.048912,0.558621,0.165517,0.612645,0.006195,0.133467,0.382811,0.537417,0.002932,0.051381,0.371516
2,3.600500e+10,7,6.142857,43.0,0.857143,42957.0,0.181281,0.658802,0.307320,0.230805,0.000000,0.206788,0.347197,0.548827,0.046914,0.017302,0.182737
3,3.600500e+10,6,17.166667,103.0,4.333333,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,3.600500e+10,13,10.076923,131.0,2.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [30]:

# Neighborhoods with highest concentrations

top_tracts = (
    tract_summary
    .sort_values(
        "final_sale_properties",
        ascending=False
    )
)

display(
    top_tracts[[
        "geoid",
        "final_sale_properties",
        "poverty_rate",
        "rent_burden_rate",
        "median_household_income",
        "avg_open_hpd_violations"
    ]].head(25)
)

,geoid,final_sale_properties,poverty_rate,rent_burden_rate,median_household_income,avg_open_hpd_violations
1596,3.608501e+10,30,0.016652,0.536379,107602.0,1.866667
695,3.604706e+10,29,0.057389,0.653386,111894.0,0.034483
1468,3.608109e+10,28,0.065832,0.470523,98921.0,0.571429
1191,3.608103e+10,27,NaN,NaN,NaN,27.851852
1637,3.608502e+10,25,NaN,NaN,NaN,12.800000
1579,3.608500e+10,25,0.276766,0.400458,61591.0,20.160000
270,3.600505e+10,25,NaN,NaN,NaN,4.080000
1486,3.608110e+10,25,0.233870,0.779438,70795.0,12.360000
1577,3.608500e+10,24,0.241275,0.657925,47755.0,3.666667
1500,3.608110e+10,23,NaN,NaN,NaN,16.391304


In [32]:

# Correlation analysis

corr_cols = [
    "final_sale_properties",
    "avg_open_hpd_violations",
    "poverty_rate",
    "rent_burden_rate",
    "severe_rent_burden_rate",
    "homeownership_rate",
    "vacancy_rate",
    "age_65_plus_rate",
    "black_rate",
    "hispanic_rate",
    "white_nonhispanic_rate",
    "bachelors_plus_rate",
    "median_household_income"
]

corr_matrix = (
    tract_summary[corr_cols]
    .corr(numeric_only=True)
)

display(corr_matrix)

,final_sale_properties,avg_open_hpd_violations,poverty_rate,rent_burden_rate,severe_rent_burden_rate,homeownership_rate,vacancy_rate,age_65_plus_rate,black_rate,hispanic_rate,white_nonhispanic_rate,bachelors_plus_rate,median_household_income
final_sale_properties,1.000000,0.017419,0.087115,0.094668,0.078080,0.018541,-0.029586,-0.166223,0.349306,-0.009650,-0.244747,-0.214754,-0.010691
avg_open_hpd_violations,0.017419,1.000000,0.183392,0.041217,0.053847,-0.271821,-0.008709,-0.141033,0.076274,0.182626,-0.131296,-0.072524,-0.007494
poverty_rate,0.087115,0.183392,1.000000,0.314742,0.365223,-0.577354,-0.118542,-0.255040,0.102871,0.410550,-0.299184,-0.467487,-0.121614
rent_burden_rate,0.094668,0.041217,0.314742,1.000000,0.689207,-0.067290,-0.108274,-0.040564,0.063552,0.174947,-0.229676,-0.398938,-0.055600
severe_rent_burden_rate,0.078080,0.053847,0.365223,0.689207,1.000000,-0.056145,-0.090468,-0.033993,0.029778,0.155174,-0.163051,-0.370326,-0.053459
homeownership_rate,0.018541,-0.271821,-0.577354,-0.067290,-0.056145,1.000000,-0.017169,0.432723,0.044194,-0.455963,0.179063,0.109596,0.030172
vacancy_rate,-0.029586,-0.008709,-0.118542,-0.108274,-0.090468,-0.017169,1.000000,0.010090,-0.088973,-0.155488,0.213188,0.282755,0.042925
age_65_plus_rate,-0.166223,-0.141033,-0.255040,-0.040564,-0.033993,0.432723,0.010090,1.000000,-0.000887,-0.296983,0.129418,0.071663,-0.013191
black_rate,0.349306,0.076274,0.102871,0.063552,0.029778,0.044194,-0.088973,-0.000887,1.000000,-0.210891,-0.598150,-0.300449,-0.051197
hispanic_rate,-0.009650,0.182626,0.410550,0.174947,0.155174,-0.455963,-0.155488,-0.296983,-0.210891,1.000000,-0.434553,-0.455223,-0.065755


In [37]:

# Simple vulnerability index
from scipy.stats import zscore

vulnerability_vars = [
    "poverty_rate",
    "rent_burden_rate",
    "vacancy_rate",
    "age_65_plus_rate",
    "avg_open_hpd_violations"
]

for col in vulnerability_vars:
    tract_summary[f"{col}_z"] = zscore(
        tract_summary[col].fillna(tract_summary[col].median())
    )

# Reverse-code protective variables
tract_summary["income_inverse_z"] = zscore(
    -tract_summary["median_household_income"].fillna(
        tract_summary["median_household_income"].median()
    )
)

tract_summary["homeownership_inverse_z"] = zscore(
    -tract_summary["homeownership_rate"].fillna(
        tract_summary["homeownership_rate"].median()
    )
)

tract_summary["vulnerability_index"] = (
    tract_summary[[f"{c}_z" for c in vulnerability_vars]].sum(axis=1)
    + tract_summary["income_inverse_z"]
    + tract_summary["homeownership_inverse_z"]
)

display(
    tract_summary["vulnerability_index"].describe()
)

count    1.665000e+03
mean    -1.365608e-16
std      2.842117e+00
min     -7.613229e+00
25%     -1.664063e+00
50%     -2.235827e-01
75%      1.459056e+00
max      1.624830e+01
Name: vulnerability_index, dtype: float64

In [ ]:
# Most vulnerable enforcement tracts

most_vulnerable = (
    tract_summary
    .sort_values(
        "vulnerability_index",
        ascending=False
    )
)

display(
    most_vulnerable[[
        "geoid",
        "final_sale_properties",
        "vulnerability_index",
        "poverty_rate",
        "rent_burden_rate",
        "median_household_income",
        "avg_open_hpd_violations"
    ]].head(30)
)

In [ ]:
# Scatterplot — vulnerability vs enforcement concentration

import matplotlib.pyplot as plt

plt.figure(figsize=(10, 6))

plt.scatter(
    tract_summary["vulnerability_index"],
    tract_summary["final_sale_properties"],
    alpha=0.6
)

plt.xlabel("Neighborhood Vulnerability Index")
plt.ylabel("Final Sale Properties")
plt.title("Enforcement Exposure and Neighborhood Vulnerability")

plt.show()

In [ ]:
# Borough-level vulnerability comparison

borough_vulnerability = (
    master_property
    .groupby("borough_lien", as_index=False)
    .agg(
        properties=("bbl", "nunique"),
        median_income=("median_household_income", "mean"),
        poverty_rate=("poverty_rate", "mean"),
        rent_burden_rate=("rent_burden_rate", "mean"),
        avg_hpd_violations=("open_hpd_violations", "mean")
    )
)

display(borough_vulnerability)